substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/resources/NativeImageResourceFileSystem.java

```diff
@@ -0,0 +1,1217 @@
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystemException;
+import java.nio.file.FileSystemNotFoundException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.OpenOption;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.FileTime;
+import java.nio.file.attribute.UserPrincipalLookupService;
+import java.nio.file.spi.FileSystemProvider;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.regex.Pattern;
+
+import org.graalvm.collections.MapCursor;
+
+/**
+ * <p>
+ * Some parts of code from this class are copied from jdk.nio.zipfs.ZipFileSystem with small tweaks.
+ * The main reason why we cannot reuse this class (without any code copying) is that we cannot
+ * extend jdk.nio.zipfs.ZipPath which is central class in Zip file system.
+ * </p>
+ */
+public class NativeImageResourceFileSystem extends FileSystem {
+
+    private static final String GLOB_SYNTAX = "glob";
+    private static final String REGEX_SYNTAX = "regex";
+
+    private static final Set<String> supportedFileAttributeViews = Collections.unmodifiableSet(
+                    new HashSet<>(Arrays.asList("basic", "resource")));
+
+    private final Set<InputStream> streams = Collections.synchronizedSet(new HashSet<>());
+    private final Set<Path> tmpPaths = Collections.synchronizedSet(new HashSet<Path>());
+
+    private final long defaultTimestamp = System.currentTimeMillis();
+
+    private final NativeImageResourceFileSystemProvider provider;
+    private final Path resourcePath;
+    private final NativeImageResourcePath root;
+    private boolean isOpen = true;
+    private final boolean useTempFile;
+    private final ReadWriteLock rwlock = new ReentrantReadWriteLock();
+
+    private static final byte[] ROOT_PATH = new byte[]{'/'};
+    private final IndexNode lookupKey = new IndexNode(null, true);
+    private final LinkedHashMap<IndexNode, IndexNode> inodes = new LinkedHashMap<>(10);
+
+    public NativeImageResourceFileSystem(NativeImageResourceFileSystemProvider provider, Path resourcePath, Map<String, ?> env) {
+        this.provider = provider;
+        this.resourcePath = resourcePath;
+        this.root = new NativeImageResourcePath(this, new byte[]{'/'});
+        if (!isTrue(env, "create")) {
+            throw new FileSystemNotFoundException(resourcePath.toString());
+        }
+        this.useTempFile = isTrue(env, "useTempFile");
```

Comment:
If I understood it correctly, this option allows us to create temporary files when modifying a file system entry? 
Would such a flag be required? This seems to only be used when attempting to create new output streams, but it may help reduce complexity if we remove it and keep everything in memory

Reply:
Correct. No, it's not required, we can get rid of this flag and simplify the operation of creating a new output stream and keep everything in memory.